src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+                whenFalseBuilder.Add(pcd);
+            }
         }
-        public override int GetHashCode()
-        {
-            return this.Input.GetHashCode() ^ (this.Symbol?.GetHashCode() ?? 0);
-        }
-        public static bool operator ==(BoundDagEvaluation left, BoundDagEvaluation right)
-        {
-            return (left == (object)null) ? right == (object)null : left.Equals(right);
-        }
-        public static bool operator !=(BoundDagEvaluation left, BoundDagEvaluation right)
-        {
-            return !left.Equals(right);
-        }
-    }
 
-    partial class BoundDagTemp
-    {
-        public override bool Equals(object obj) => obj is BoundDagTemp other && this.Equals(other);
-        public bool Equals(BoundDagTemp other)
+        /// <summary>
+        /// Given that the decision d has occurred and produced a true/false result,
+        /// set permitsTrue if a true decision on d would permit other to succeed.
+        /// set permitsFalse if a false decision on d would permit other to succeed.
+        /// sets killsFalseDecision when d being false means other has been proven true
+        /// </summary>
+        private void CheckConsistentDecision(
+            BoundDagDecision d,
+            BoundDagDecision other,
+            out bool permitsTrue,
+            out bool permitsFalse,
+            out bool killsDecisionOnTrueBranch,
+            out bool killsDecisionOnFalseBranch)
         {
-            return other != (object)null && this.Type == other.Type && object.Equals(this.Source, other.Source) && this.Index == other.Index;
+            // innocent until proven guilty
+            permitsTrue = true;
+            permitsFalse = true;
+            killsDecisionOnTrueBranch = false;
+            killsDecisionOnFalseBranch = false;
+
+            // if decisions test unrelated things, there is no implication from one to the other
+            if (d.Input != other.Input)
+            {
+                return;
+            }
+
+            // a test cannot be both true and false
+            if (d == other)
+            {
+                permitsFalse = false;
+                killsDecisionOnTrueBranch = true;
+                return;
+            }
+
+            switch (d)
+            {
+                case BoundNonNullDecision n1:
+                    switch (other)
+                    {
+                        case BoundValueDecision v2:
+                            if (v2.Value == ConstantValue.Null)
+                            {
+                                // once v!=null is false, we know v==null is true and do not need to test it
+                                permitsTrue = false;
+                                killsDecisionOnFalseBranch = true;
+                            }
+                            else
+                            {
+                                // Given that v!=null fails, v==K might succeed
+                                permitsFalse = false;
```

Comment:
>permitsFalse = false; [](start = 32, length = 21)

The comment says "v==K might succeed", but it looks like we are marking is as not successful. Should the comment be adjusted?

Reply:
Addressed in https://github.com/dotnet/roslyn/pull/24059/commits/114a56e9077ed03f08443889b69cae6972a4acc6 of https://github.com/dotnet/roslyn/pull/24059